staging/src/k8s.io/kubectl/pkg/cmd/run/run.go

```diff
@@ -381,17 +381,21 @@ func (o *RunOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []string) e
 		}
 
 		var pod *corev1.Pod
-		leaveStdinOpen := o.LeaveStdinOpen
-		waitForExitCode := !leaveStdinOpen && restartPolicy == corev1.RestartPolicyNever
+		waitForExitCode := !o.LeaveStdinOpen && (restartPolicy == corev1.RestartPolicyNever || restartPolicy == corev1.RestartPolicyOnFailure)
 		if waitForExitCode {
-			pod, err = waitForPod(clientset.CoreV1(), attachablePod.Namespace, attachablePod.Name, opts.GetPodTimeout, podCompleted)
+			// we need different exit condition depending on resatart policy
+			// for Never it can either fail or succeed, for OnFailure only
+			// success matters
+			exitCondition := podCompleted
+			if restartPolicy == corev1.RestartPolicyOnFailure {
+				exitCondition = podSucceeded
```

Comment:
Yeah - I think the failure from: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/90448/pull-kubernetes-e2e-gce/1253642598376017920/ is actually a valid one.
Basically, in that test the pod will never succeed, because it always "exit 42".

Not sure if that means this condition is not the right one, or whether the test itself should be changed.

Reply:
I think the test is wrong, we can't expect this to succeed if the command is always failing causing the pod to be restarted over and over again. That's why I'm updating the test to expect a time out failure.